mesa/model.py

```diff
@@ -94,16 +99,51 @@ def agents(self, agents: Any) -> None:
             "Please adjust your code to use a different attribute name for custom agent storage."
         )
 
-        self._agents = agents
-
     @property
     def agent_types(self) -> list[type]:
         """Return a list of different agent types."""
-        return list(self.agents_.keys())
+        return list(self.agents_by_type.keys())
 
     def get_agents_of_type(self, agenttype: type[Agent]) -> AgentSet:
```

Comment:
I think we can deprecate this function and just let users do `self.agents_by_type[agenttype]`.

Reply:
I have been going back and forth on this. It all depends on which attributes of the agent registration stuff become part of the public api and which ones we want to keep protected.

Reply:
Another option would be to make a fast path in `self.agents.select(type=agenttype)`

Reply:
That would be tricky because it would involve messing with the internal structure of AgentSet.